src/Collection.php

```diff
@@ -228,22 +225,18 @@ public function aggregate(array $pipeline, array $options = [])
             $options['readPreference'] = $this->readPreference;
         }
 
-        $server = select_server($this->manager, $options);
-
-        /* If a write stage is being used with a read preference (explicit or
-         * inherited), check that the wire version supports it. If not, force a
-         * primary read preference and select a new server if necessary. */
+        /* If a write stage is being used with a non-primary read preference and
+         * at least one server is pre-5.0, force a primary read preference. */
         if (
             $hasWriteStage && isset($options['readPreference']) &&
-            ! server_supports_feature($server, self::$wireVersionForSecondarySupportsWriteStage)
+            $options['readPreference']->getMode() !== ReadPreference::RP_PRIMARY &&
+            ! all_servers_support_write_stage_on_secondary($this->manager)
```

Comment:
Since the whole point of the spec change ([DRIVERS-1969](https://jira.mongodb.org/browse/DRIVERS-1969)) is to ensure we fall back to a primary read preference, we only need to bother with `all_servers_support_write_stage_on_secondary` if a non-primary read preference was specified (or inherited).